core/api/src/main/scala/org/virtuslab/ideprobe/jsonrpc/logging/RequestResponseLogger.scala

```diff
@@ -0,0 +1,76 @@
+package org.virtuslab.ideprobe.jsonrpc.logging
+
+import java.time.ZonedDateTime
+import java.util.concurrent.{ScheduledFuture, ScheduledThreadPoolExecutor, TimeUnit}
+
+class RequestResponseLogger {
+
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private val ticker: Runnable = () => flush()
+  private var task: Option[ScheduledFuture[_]] = None
+
+  private sealed trait LogEntry
+  private case class Request(message: String) extends LogEntry
+  private case class Response(message: String) extends LogEntry
+  private case class RequestAndResponse(request: String, response: String) extends LogEntry
+
+  private var buffer: List[(LogEntry, ZonedDateTime)] = List.empty
+
+  def logRequest(request: String): Unit = buffer.synchronized {
+    task.foreach(_.cancel(true))
+    buffer match {
+      case (head @ (RequestAndResponse(`request`, _), _)) :: tail =>
+        buffer = (Request(request), ZonedDateTime.now()) :: head :: tail
+      case _ =>
+        flush()
+        buffer = (Request(request), ZonedDateTime.now()) :: Nil
+    }
+    task = Option(ex.schedule(ticker, 1, TimeUnit.SECONDS))
+  }
+
+  def logResponse(response: String): Unit = buffer.synchronized {
+    task.foreach(_.cancel(true))
+    buffer match {
+      case (Request(request), timestamp) :: Nil =>
+        buffer = (RequestAndResponse(request, response), timestamp) :: Nil
+      case (Request(ultimate), timestamp) :: (penultimate @ (RequestAndResponse(_, `response`), _)) :: tail =>
+        buffer = (RequestAndResponse(ultimate, response), timestamp) :: penultimate :: tail
+      case (Request(request), timestamp) :: tail =>
+        buffer = tail
+        flush()
+        buffer = (RequestAndResponse(request, response), timestamp) :: Nil
+      case _ =>
+        buffer = (Response(response), ZonedDateTime.now()) :: buffer
+        flush()
+    }
+    task = Option(ex.schedule(ticker, 1, TimeUnit.SECONDS))
+  }
+
+  private def flush(): Unit = buffer.synchronized {
+    val now = ZonedDateTime.now()
+    val firstMessageTimestamp = buffer.reverse.headOption.map(_._2).getOrElse(now)
```

Comment:
wouldn't this mean an empty list? How about early return? If list is empty - do nothing, else take last element, and don't bother handling options. Currently it feels like this `now` in firstMessageTimestamp is used for anything

Reply:
Yeah, it is in fact unused when the list is empty. Fixed.